src/preamble.js

```diff
@@ -2318,19 +2318,38 @@ function integrateWasmJS() {
     // later), so we save Module and check it later.
     var trueModule = Module;
 #endif
-    getBinaryPromise().then(function(binary) {
-      return WebAssembly.instantiate(binary, info)
-    }).then(function(output) {
+    function receiveInstantiatedSource(output) {
+      // 'output' is a WebAssemblyInstantiatedSource object which has both the module and instance.
       // receiveInstance() will swap in the exports (to Module.asm) so they can be called
 #if ASSERTIONS
       assert(Module === trueModule, 'the Module object should not be replaced during async compilation - perhaps the order of HTML elements is wrong?');
       trueModule = null;
 #endif
       receiveInstance(output['instance']);
-    }).catch(function(reason) {
-      Module['printErr']('failed to asynchronously prepare wasm: ' + reason);
-      abort(reason);
-    });
+    }
+    function instantiateArrayBuffer(receiver) {
+      getBinaryPromise().then(function(binary) {
+        return WebAssembly.instantiate(binary, info);
+      }).then(receiver).catch(function(reason) {
+        Module['printErr']('failed to asynchronously prepare wasm: ' + reason);
+        abort(reason);
+      });
+    }
+    // Prefer streaming instantiation if available.
+    if (!Module['wasmBinary'] && typeof WebAssembly.instantiateStreaming === 'function') {
+      WebAssembly.instantiateStreaming(fetch(wasmBinaryFile, { credentials: 'same-origin' }), info)
```

Comment:
The indentation made this a little hard for me to read (to see where the `.`s go etc.). How about

````
WebAssembly.instantiateStreaming(
  fetch(wasmBinaryFile, { credentials: 'same-origin' }), info)
    .then(receiveInstantiatedSource)
    .catch(function(reason) {
      // We expect the most common failure cause to be a bad MIME type for the binary,
      // in which case falling back to ArrayBuffer instantiation should work.
      Module['printErr']('wasm streaming compile failed: ' + reason);
      Module['printErr']('falling back to ArrayBuffer instantiation');
      instantiateArrayBuffer(receiveInstantiatedSource);
    })
);
````

Also I had to add a `)` so I think one might have been missing before.

Reply:
Actually, that indentation isn't right because the `.then` and `.catch` are not attached to the promise returned by `fetch`, they go with the return of `instantiateStreaming`. The `fetch`'s promise is the first argument to `instantiateStreaming`. (That is also the reason why the number of parens is correct). I did make the `.catch` line up with the `.then` though, agreed that's better.

Reply:
cool, looks nice.